### PR TITLE
Homework 3

### PR DESCRIPTION
**Pull request by 18Jura12**
No scripts, made some Controller adjustments.
First run dbinitializers for both databases.

### CODE REVIEW

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+				SelectionCoefficient: bet.SelectionCoefficient,
+				Payment:              bet.Payment,
+			}
+
+			err := h.betRepository.InsertCalcBet(ctx, domainBet)
+			if err != nil {
+				log.Println("Failed to insert bet, error: ", err)
+				continue
+			}
+			log.Println("Inserted bet, betId: ", domainBet.Id, ", selectionId: ", domainBet.SelectionId)
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: What if bet already exists? You will log `uniqueViolationError` (I have never used `sqllite`, but it probably works like that).

In that case you should not log anything, or you could inform application developer through logs about that occasion. (With different log message, something user friendly).
Or maybe you could update existing bet in the DB with new coefficient.

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+	go func() {
+		defer close(betsCalculated)
+
+		for eventUpdate := range eventUpdates {
+			log.Println("Processing event update, Id:", eventUpdate.Id)
+
+			domainBets, exists, err := h.betRepository.GetCalcBetsBySelectionID(ctx, eventUpdate.Id)
+			if err != nil {
+				log.Println("Failed to fetch a bet which should be updated, error: ", err)
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: `bets`

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+				continue
+			}
+			if !exists {
+				log.Println("A bet which should be updated does not exist, selection id: ", eventUpdate.Id)
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: This message assumes that there should be bets for the `eventUpdate.Id`.

It is completely fine that there are no bets for the mentioned event.
Something like: `event update <x> references 0 bets` could be more appropriate.

File: lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/betConsumer.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "bet consumer failed to consume messages")
+	}
+
+	bets := make(chan models.Bet)
+	go func() {
+		defer close(bets)
+		for msg := range msgs {
+			var bet models.Bet
+			err := json.Unmarshal(msg.Body, &bet)
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/betConsumer.go:58]: What happens if this fails?

File: lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/eventUpdateConsumer.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	eventUpdates := make(chan models.EventUpdate)
+	go func() {
+
+		defer close(eventUpdates)
+		for msg := range msgs {
+			var eventUpdate models.EventUpdate
+			err := json.Unmarshal(msg.Body, &eventUpdate)
+			if err != nil {
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/eventUpdateConsumer.go:60]: Here also?

File: lecture_3/03_project/calculator/internal/engine/handler/calcbetRepository.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+	"context"
+	domainmodels "github.com/superbet-group/code-cadets-2021/lecture_3/03_project/calculator/internal/domain/models"
+)
+
+type CalcBetRepository interface {
+	InsertCalcBet(ctx context.Context, bet domainmodels.Bet) error
+	UpdateCalcBet(ctx context.Context, bet domainmodels.Bet) error
+	GetCalcBetByID(ctx context.Context, id string) (domainmodels.Bet, bool, error)
+	GetCalcBetsBySelectionID(ctx context.Context, selectionId string) ([]domainmodels.Bet, bool, error)
+}
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/calcbetRepository.go]: Interface should always define a minimal set of methods used by its owner.

Comment (svennjegacAxilis) [reply lecture_3/03_project/calculator/internal/engine/handler/calcbetRepository.go]: (Here, methods which are unused exist)

File: lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/betConsumer.go (hunk 4 above)
Comment (18Jura12) [reply lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/betConsumer.go:58]: If it fails, we are notified and the go routine keeps on reading the rest of the messages. I don't see a problem.
I didn't want to abort a go routine.

Comment (18Jura12) [reply lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/betConsumer.go:58]: I can, however, propagate it, if that will be better.

Comment (svennjegacAxilis) [reply lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/betConsumer.go:58]: You resolved the issue I was aiming for. You added `continue` in the last PR :)

Before, you logged that error occurred and produced "gibberish" from `var bet models.Bet` into the `bets` channel.

So it works fine now. Please add `continue` to the `eventUpdateConsumer` as well.

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+			}
+
+			err := h.betRepository.InsertCalcBet(ctx, domainBet)
+			if err == sqlite3.ErrConstraint {
+				log.Println("bet already exists in calc_bets")
+				continue
+			} else if err != nil {
+				log.Println("Failed to insert bet, error: ", err)
+				continue
+			}
```
Comment (svennjegacAxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go:40]: This handler should not be aware that `sqlite3` is used under the hood. That information should be hidden and exposed differently.

E.g. you could change the signature of your `repository` method to:
`InsertCalcBet(ctx context.Context, bet domainmodels.Bet) (bool, error)`.
And then, through `bool` return value you can signal if it was actually inserted or not.

**Review by svennjegacAxilis — APPROVED** [review]
Left 1 comment, but you don't have to implement it, if you don't have enough time (so I'm approving this PR).

At least think about what I said and consider that in your future PRs (through the Go School, and through all of the other projects as well).

All in all, pretty good homework. I didn't have any major comments 👍

**Pull request merged**